### ISSUE TITLE
About hyperparameter tuning guide

### ISSUE THREAD

**Issue by hyebinyoo** [ultralytics/ultralytics#21919]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi. I followed the hyperparameter tuning guide in the following link: [https://docs.ultralytics.com/ko/guides/hyperparameter-tuning]

But I cannot see an upward tendency like the best_fitness.png picture you put as an example.

I couldn't see the upward trend with my custom weight file and custom dataset, so I tried to copy your code.

Like this:
`
def finetune(args):
    model = YOLO("yolo11n.pt")

    search_space = {
        # LR
        "lr0": [1e-5, 1e-1],
        "lrf": [0.01, 1.0],
        "momentum": [0.6, 0.98],
        "weight_decay": [0.0, 0.001],

        # Warmup
        "warmup_epochs": [0.0, 5.0],
        "warmup_momentum": [0.0, 0.95],

        # Loss
        "box": [0.02, 0.2],
        "cls": [0.2, 4.0],

        # Augmentations
        "hsv_h": [0.0, 0.1],
        "hsv_s": [0.0, 0.9],
        "hsv_v": [0.0, 0.9],
        "degrees": [0.0, 45.0],
        "translate": [0.0, 0.9],
        "scale": [0.0, 0.9],
        "shear": [0.0, 10.0],
        "perspective": [0.0, 0.001],
        "flipud": [0.0, 1.0],
        "fliplr": [0.0, 1.0],
        "mosaic": [0.0, 1.0],
        "mixup": [0.0, 1.0],
        "copy_paste": [0.0, 1.0],
    }

    model.tune(
        data=args.data_file,
        epochs=30,
        iterations=300,
        optimizer="AdamW",
        space=search_space,
        batch=4,
        cache="disk",
        plots=False, 
        save=True, 
        val=True,
        patience=0,
        project = "runs/tune",
        name=f"{args.tune_path}_coco8_adamw",
        resume=True
    )

    with open(f"runs/tune/{args.tune_path}_coco8_adamw/best_hyperparameters.yaml", "r", encoding="utf-8") as f:
        lr_params = yaml.safe_load(f)

    model.train(
        data=args.data_file,
        epochs=100,
        imgsz=640,
        batch=4,
        optimizer="AdamW",
        device=0,
        save=True,
        verbose=True,
        patience=30,
        **lr_params,
        project = "runs/tune_train",
        name="final_train_AdamW"
    )

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="train config")
    parser.add_argument("--data_file", help="data yaml file")  # coco8.yaml
    parser.add_argument("--tune_path", default="auto_tune")
    args = parser.parse_args()
    finetune(args)
`

But the graph didn't show any tendency, not to the right:

<img width="2000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/bc1c4b02-1f9f-45b5-91d2-ddf8506d91ef" />

Can you guys tell me the "that code" that made that right-up graph?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @hyebinyoo, thanks for opening this issue and for working with Ultralytics 🚀! This is an automated response to help get things moving quickly — an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

To help us examine your hyperparameter tuning behavior and plots, please include:
-Exact ultralytics version and install method
-Full tune/train commands or script snippets you ran
-Complete console logs from tune and train runs
-Your data YAML, number of images per split, and example images
-The resulting runs/tune/*/results.csv and best_hyperparameters.yaml
-Any seeds you set and whether resume was used across attempts
-A screenshot of the best_fitness and fitness curves

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

-**Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-**Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-**Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-**Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the clear repro and plots. The “up‑and‑right” fitness chart is produced from your tuning CSV by the plotting helper below; it saves next to the CSV as tune_fitness.png as documented in the [plot_tune_results reference](https://docs.ultralytics.com/reference/utils/plotting/):

```python
from ultralytics.utils.plotting import plot_tune_results
plot_tune_results("path/to/your/tune_results.csv")
```

Trend isn’t guaranteed—on small/custom datasets it can plateau or oscillate—so for a fair run please start fresh (`resume=False`), define bounds as tuples not lists (or omit `space` to use the default search space from the [Tuner class reference](https://docs.ultralytics.com/reference/engine/tuner/)), and consider `use_ray=True` to enable ASHA early‑stopping as shown in the [Ray Tune integration guide](https://docs.ultralytics.com/integrations/ray-tune/):

```python
from ultralytics import YOLO
YOLO("yolo11n.pt").tune(data="coco8.yaml", use_ray=True, iterations=50, epochs=30, resume=False)
```

If you’d like us to check what’s happening in your run, please attach the text of your tune_results.csv from the tuning directory rather than screenshots.

**Comment by hyebinyoo** [comment]
Thank yo for your answer.

I thought it was guaranteed to be rightward like your drawings in the hyperparameter tuning [https://docs.ultralytics.com/ko/guides/hyperparameter-tuning/](url) link, because it says it's done with GA techniques. 

What I'm curious about is

1. How do I make it look like this right-wing graph of you guys?
(Which hyperparameter was used as which minmax value, what was used as the dataset, etc.)
In that link guide, you used AdamW as optimizer, and dataset as coco8. Then, it's guided as if there's an upward trend.
I'd like to know the code that makes you look like right-wingers. If it's hard for you to tell me, you can tell me it's hard.


2. Why isn't any trend guaranteed when using gene mutation techniques?
I agree that this can happen if there are too few datasets. However, the fact that there are enough datasets and no trends are seen is questionable.

**Comment by Y-T-G** [comment]
What's the size of your dataset?

**Comment by Y-T-G** [comment]
What's your version of Ultralytics?

**Comment by hyebinyoo** [comment]
I used the coco8 in your example code. The coco8 with 4 train and 4 val

The version of ultralytics is 8.3.175.

**Comment by Y-T-G** [comment]
`coco8.yaml` is a toy dataset. The model wouldn't learn much from it. You need to use a larger dataset to see the effect.

**Comment by hyebinyoo** [comment]
That's why I wonder what you guys use in that guide link.

I referenced your example code in hyperaparameter tuning guide link:

```
from ultralytics import YOLO

# Initialize the YOLO model
model = YOLO("yolo11n.pt")

# Define search space
search_space = {
    "lr0": (1e-5, 1e-1),
    "degrees": (0.0, 45.0),
}

# Tune hyperparameters on COCO8 for 30 epochs
model.tune(
    data="coco8.yaml",
    epochs=30,
    iterations=300,
    optimizer="AdamW",
    space=search_space,
    plots=False,
    save=False,
    val=False,
)

```

 because even though I used my custom dataset with enough size, there was no upward trend.


How did you tune a model and what did you use dataset to gain the graph that has right-up trend shape?

<img width="655" height="379" alt="Image" src="https://github.com/user-attachments/assets/0dfefda6-080a-445b-993a-da999d41c1bc" />

**Comment by Y-T-G** [comment]
It's probably `coco.yaml`

**Comment by hyebinyoo** [comment]
Coco.yaml is about 110,000 copies, but you guys used this?

**Comment by glenn-jocher** [comment]
Yes—those illustrative curves were generated on a COCO‑size dataset (`coco.yaml`) with a larger tuning budget; on `coco8` the per‑trial fitness is too noisy to show a clear rise. To reproduce on your side, run a fresh tune on COCO with more iterations/epochs and a fixed seed (the running best will trend upward even though individual trials bounce):

```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
model.tune(
    data="coco.yaml",
    iterations=80,
    epochs=100,
    resume=False,
    seed=0,
)
```

If you want a strictly non‑decreasing curve, plot the running best from your `tune_results.csv`:

```python
import pandas as pd
df = pd.read_csv("runs/detect/tune/tune_results.csv")
best = df["fitness"].cummax()
best.to_frame("best_fitness").plot()
```

Background on why GA sampling doesn’t guarantee per‑iteration improvement (only best‑so‑far typically rises) is covered in our [hyperparameter tuning overview](https://www.ultralytics.com/glossary/hyperparameter-tuning).

**Comment by hyebinyoo** [comment]
[Hyperparameter tuning overview](https://www.ultralytics.com/ko/glossary/hyperparameter-tuning) doesn't seem to indicate why GA has no per-iteration implication guarante.

So you guys trained the entire coco dataset (11.8 million copies) for 80epoch using yolo11n.pt ?

But you guys look like you stopped at 558 of 900 iteration

I understand that you draw a graph using the tune_result.csv file.
But that's not what I'm curious about. What should I do to make it look like the top-right graph that you guys posted on the guide

GA shows too much improvement in the graph, contrary to your statement that the per-iteration imposition is not visible.

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—quick clarifications and a reproducible recipe:

- The figure in the guide was run on full COCO (`coco.yaml`, ~118k train images, not 11.8M) with YOLO11n, AdamW, default search space, `seed=0`, and a larger budget (e.g., hundreds of iterations; your “558/900 iterations” observation matches that). The curve shown is the running best fitness, not raw per‑trial fitness, which explains the steady up‑and‑right shape. See the `tune()` behavior in the [model.tune reference](https://docs.ultralytics.com/reference/engine/model/).

- Genetic mutation explores new candidates each iteration, so individual trial fitness will bounce due to stochastic training and validation noise; only the best‑so‑far is expected to be non‑decreasing. Ray’s ASHA scheduler further biases progress by early‑stopping weak trials; details are in the [Ray Tune utility reference](https://docs.ultralytics.com/reference/utils/tuner/).

To reproduce a similar trend on your side, use COCO and a larger budget with Ray + ASHA:

```python
from ray import tune
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
model.tune(
    data="coco.yaml",
    use_ray=True,
    iterations=200,      # number of trials (max_samples)
    epochs=50,           # per-trial epochs
    seed=0,
    resume=False,
    grace_period=10,     # ASHA warmup
    space={"lr0": tune.uniform(1e-5, 1e-1), "degrees": tune.uniform(0.0, 45.0)},
)
```

On `coco8` the signal is too noisy to show a smooth rise; use your larger dataset or COCO for the “up‑and‑right” running‑best curve.

**Comment by hyebinyoo** [comment]
Thank you for your answer! I'll close the issue.

**Comment by glenn-jocher** [comment]
Glad it’s resolved—if you open a follow‑up, please include `yolo checks` (or `!yolo checks` in notebooks) and the text of `runs/detect/tune/tune_results.csv` for faster triage.
